tree_assume/src/typechecker.rs

```diff
@@ -0,0 +1,301 @@
+                };
+                let mut new_branches = vec![];
+                let mut res_type = None;
+                for branch in branches {
+                    let (bty, new_branch) =
+                        self.add_arg_types_to_expr(branch.clone(), arg_ty.clone());
+                    new_branches.push(new_branch);
+                    res_type = match res_type {
+                        Some(t) => {
+                            assert_eq!(t, bty, "Expected all branches to have the same type");
+                            Some(t)
+                        }
+                        None => Some(bty),
+                    };
+                }
+                (
+                    res_type.unwrap(),
+                    RcExpr::new(Expr::Switch(new_integer, new_branches)),
+                )
+            }
+            Expr::If(pred, then, else_branch) => {
+                let (pty, new_pred) = self.add_arg_types_to_expr(pred.clone(), arg_ty.clone());
+                let Type::Base(BaseType::BoolT) = pty else {
+                    panic!("Expected bool type. Got {:?}", pty)
+                };
+                let (tty, new_then) = self.add_arg_types_to_expr(then.clone(), arg_ty.clone());
+                let (ety, new_else) =
+                    self.add_arg_types_to_expr(else_branch.clone(), arg_ty.clone());
+                assert_eq!(
+                    tty, ety,
+                    "Expected then and else types to be the same. Got {:?} and {:?}",
+                    tty, ety
+                );
+                (tty, RcExpr::new(Expr::If(new_pred, new_then, new_else)))
+            }
+            Expr::Let(input, body) => {
+                let (ity, new_input) = self.add_arg_types_to_expr(input.clone(), arg_ty.clone());
+                let (bty, new_body) = self.add_arg_types_to_expr(body.clone(), ity);
+                (bty, RcExpr::new(Expr::Let(new_input, new_body)))
+            }
+            Expr::DoWhile(inputs, pred_and_outputs) => {
+                let (ity, new_inputs) = self.add_arg_types_to_expr(inputs.clone(), arg_ty.clone());
+                let Type::TupleT(in_tys) = ity.clone() else {
+                    panic!("Expected tuple type. Got {:?}", ity)
+                };
+                let (pty, new_pred_and_outputs) =
+                    self.add_arg_types_to_expr(pred_and_outputs.clone(), ity);
+                let Type::TupleT(out_tys) = pty else {
+                    panic!("Expected tuple type. Got {:?}", pty)
+                };
+                assert_eq!(out_tys[0], Type::Base(BaseType::BoolT));
+                assert_eq!(out_tys[1..], in_tys);
+                (
+                    Type::TupleT(out_tys[1..].to_vec()),
+                    RcExpr::new(Expr::DoWhile(new_inputs, new_pred_and_outputs)),
+                )
+            }
+            // Replace the argument type with the new type
+            Expr::Arg(_ignored_ty) => (arg_ty.clone(), Rc::new(Expr::Arg(arg_ty))),
```

Comment:
why do we ignore the type?

Reply:
no longer ignored- does type checking, and replaces unknowns